power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -57,22 +59,94 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
     });
 }
 
-template <symmetry_tag sym>
-inline void calculate_source_result(IdxRange const& sources, Idx bus_number, YBus<sym> const& y_bus,
-                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output) {
+// Calculates current and power injection of source i for multiple symmetric sources at a node.
+// The current injection of source i to the bus in phase space is:
+//     i_inj_i = (y_ref_i * z_ref_t) [ (u_ref_i * y_ref_t - i_ref_t) + i_inj_t]
+// Phase space is used because y_ref is a DoubleComplex in the symmetric case.
+// The equation is organized in such a way to avoid the numerical instability induced in the substraction if y_ref_i
+// is large.
+inline void calculate_multiple_source_result(IdxRange const& sources, YBus<symmetric_t> const& y_bus,
+                                             PowerFlowInput<symmetric_t> const& input,
+                                             ComplexValue<symmetric_t> const& i_inj_t,
+                                             SolverOutput<symmetric_t>& output, Idx const& bus_number) {
+    ComplexVector const y_ref = y_bus.math_model_param().source_param;
+    DoubleComplex const y_ref_t =
+        std::accumulate(sources.begin(), sources.end(), DoubleComplex{},
+                        [&](DoubleComplex sum, Idx const source) { return y_ref[source] + sum; });
+    DoubleComplex const i_ref_t =
+        std::accumulate(sources.begin(), sources.end(), DoubleComplex{}, [&](DoubleComplex sum, Idx const source) {
+            return input.source[source] * y_ref[source] + sum;
+        });
+    for (Idx const source : sources) {
+        DoubleComplex const y_ref_i_over_y_ref_t = y_ref[source] / y_ref_t;
```

Comment:
Please cache the inverse `1.0 / y_ref_t` as this is inside a `for` loop

Reply:
Done.